### ISSUE TITLE
Investigate possible memory exhaustion errors on action failures.

### ISSUE THREAD

**Issue by james-allan** [woocommerce/action-scheduler#379]
I've just discovered the following memory exhaustion errors on one of my local installs running WC Subscriptions and AS 3.0.

```
PHP Fatal error:  Out of memory (allocated 100663296) (tried to allocate 28672 bytes) in /app/public/wp-includes/wp-db.php on line 2007
PHP Stack trace:
PHP   1. {main}() /app/public/wp-admin/admin-ajax.php:0
PHP   2. do_action() /app/public/wp-admin/admin-ajax.php:173
PHP   3. WP_Hook->do_action() /app/public/wp-includes/plugin.php:465
PHP   4. WP_Hook->apply_filters() /app/public/wp-includes/class-wp-hook.php:310
PHP   5. ActionScheduler_AsyncRequest_QueueRunner->maybe_handle() /app/public/wp-includes/class-wp-hook.php:286
PHP   6. ActionScheduler_AsyncRequest_QueueRunner->handle() /app/public/wp-content/plugins/woocommerce/includes/libraries/wp-async-request.php:147
PHP   7. do_action() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/libraries/action-scheduler/classes/ActionScheduler_AsyncRequest_QueueRunner.php:52
PHP   8. WP_Hook->do_action() /app/public/wp-includes/plugin.php:465
PHP   9. WP_Hook->apply_filters() /app/public/wp-includes/class-wp-hook.php:310
PHP  10. ActionScheduler_QueueRunner->run() /app/public/wp-includes/class-wp-hook.php:286
PHP  11. ActionScheduler_QueueRunner->do_batch() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/libraries/action-scheduler/classes/ActionScheduler_QueueRunner.php:120
PHP  12. ActionScheduler_QueueRunner->process_action() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/libraries/action-scheduler/classes/ActionScheduler_QueueRunner.php:150
PHP  13. ActionScheduler_Action->execute() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/libraries/action-scheduler/classes/abstracts/ActionScheduler_Abstract_QueueRunner.php:61
PHP  14. do_action_ref_array() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/libraries/action-scheduler/classes/actions/ActionScheduler_Action.php:22
PHP  15. WP_Hook->do_action() /app/public/wp-includes/plugin.php:531
PHP  16. WP_Hook->apply_filters() /app/public/wp-includes/class-wp-hook.php:310
PHP  17. WC_Subscriptions_Manager::prepare_renewal() /app/public/wp-includes/class-wp-hook.php:286
PHP  18. WC_Subscriptions_Manager::process_renewal() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/class-wc-subscriptions-manager.php:89
PHP  19. WC_Subscription->update_status() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/class-wc-subscriptions-manager.php:113
PHP  20. WC_Subscription->save() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/class-wc-subscription.php:513
PHP  21. WC_Subscription->status_transition() /app/public/wp-content/plugins/woocommerce/includes/class-wc-order.php:221
PHP  22. do_action() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/class-wc-subscription.php:554
PHP  23. WP_Hook->do_action() /app/public/wp-includes/plugin.php:465
PHP  24. WP_Hook->apply_filters() /app/public/wp-includes/class-wp-hook.php:310
PHP  25. wcs_maybe_make_user_inactive_for() /app/public/wp-includes/class-wp-hook.php:286
PHP  26. wcs_maybe_make_user_inactive() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/wcs-user-functions.php:52
PHP  27. wcs_user_has_subscription() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/wcs-user-functions.php:39
PHP  28. wcs_get_users_subscriptions() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/wcs-user-functions.php:135
PHP  29. wcs_get_subscription() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/wcs-user-functions.php:192
PHP  30. WC_Order_Factory::get_order() /app/public/wp-content/plugins/woocommerce-subscriptions/wcs-functions.php:84
PHP  31. WC_Subscription->__construct() /app/public/wp-content/plugins/woocommerce/includes/class-wc-order-factory.php:47
PHP  32. WC_Subscription->__construct() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/class-wc-subscription.php:117
PHP  33. WC_Data_Store->read() /app/public/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-order.php:111
PHP  34. WCS_Subscription_Data_Store_CPT->read() /app/public/wp-content/plugins/woocommerce/includes/class-wc-data-store.php:159
PHP  35. WCS_Subscription_Data_Store_CPT->read_order_data() /app/public/wp-content/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php:113
PHP  36. WCS_Subscription_Data_Store_CPT->read_order_data() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/data-stores/class-wcs-subscription-data-store-cpt.php:134
PHP  37. WCS_Subscription_Data_Store_CPT->read_order_data() /app/public/wp-content/plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:97
PHP  38. get_post_meta() /app/public/wp-content/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php:277
PHP  39. get_metadata() /app/public/wp-includes/post.php:2032
PHP  40. update_meta_cache() /app/public/wp-includes/meta.php:525
PHP  41. shutdown_action_hook() /app/public/wp-includes/load.php:0
PHP  42. do_action() /app/public/wp-includes/load.php:954
PHP  43. WP_Hook->do_action() /app/public/wp-includes/plugin.php:465
PHP  44. WP_Hook->apply_filters() /app/public/wp-includes/class-wp-hook.php:310
PHP  45. ActionScheduler_FatalErrorMonitor->handle_unexpected_shutdown() /app/public/wp-includes/class-wp-hook.php:286
PHP  46. ActionScheduler_DBStore->mark_failure() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/libraries/action-scheduler/classes/ActionScheduler_FatalErrorMonitor.php:48
PHP  47. wpdb->update() /app/public/wp-content/plugins/woocommerce-subscriptions/includes/libraries/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php:726
PHP  48. wpdb->process_fields() /app/public/wp-includes/wp-db.php:2219
PHP  49. wpdb->process_field_charsets() /app/public/wp-includes/wp-db.php:2331
PHP  50. wpdb->get_col_charset() /app/public/wp-includes/wp-db.php:2403
PHP  51. wpdb->get_table_charset() /app/public/wp-includes/wp-db.php:2760
PHP  52. wpdb->get_results() /app/public/wp-includes/wp-db.php:2657
PHP  53. wpdb->query() /app/public/wp-includes/wp-db.php:2579
PHP  54. wpdb->_do_query() /app/public/wp-includes/wp-db.php:1895
PHP  55. mysqli_query() /app/public/wp-includes/wp-db.php:2007
```

Following the stack trace, it appears to be an unexpected shutdown while processing a renewal order leads to the `ActionScheduler_DBStore->mark_failure()` function which then leads to some `wpdb` function calls and an eventual memory error.

**Comment by thenbrent** [comment]
Can you determine if the memory exhaustion causes the action's failure, or is it a result of it?

**Comment by james-allan** [comment]
> Can you determine if the memory exhaustion causes the action's failure, or is it a result of it?

I think the memory exhaustion causes the failure. I can see similar log entries with stacktraces very similar except they don't include the `ActionScheduler_FatalErrorMonitor->handle_unexpected_shutdown()` part. 

Basically are those traces the original error, and the AS related ones are either subsequent errors or the same one. 

So it could just be that the subscriptions related memory exhaustion is the root problem here.

**Comment by james-allan** [comment]
Closing. I haven't seen this since the initial report and following the reasons above, the exhaustions might have been coming from somewhere else.

**Comment by gencmedya** [comment]
memory problem.
